cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,278 @@
+  }
+
+  public Iterator<Message> aggregate(Iterator<Message> input) {
+    cachemap.clear();
+    if (!input.hasNext()) {
+      return input;
+    }
+
+    Map<MetricsMapKey, MetricsAggregator> aggregated = new HashMap<>();
+    String[] messageId = new String[1];
+    long currentTimestamp = System.currentTimeMillis() / 1000;
+    // MetricValues will be aggregated across windows of size "aggregationWindowSeconds".
+    // We align the current time to the end of the last widow. This is done since in most cases,
+    // it is expected that all the MetricValues were emitted very recently. The offset ensures
+    // the last window completely lies in the past and maximizes the chance that all metrics lie within it.
+    long offset = (currentTimestamp % aggregationWindowSeconds) + 1;
+
+    input.forEachRemaining(message -> {
+      // Decode the metrics.
+      // Keep track of the last message ID as TopicRelayers need it to fetch messages in the next iteration.
+      messageId[0] = message.getId();
+      decodingInputStream.reset(message.getPayload());
+      MetricValues values;
+      try {
+        values = reader.read(decoder, schema);
+      } catch (IOException e) {
+        LOG.warn("Failed to decode metric message.", e);
+        return;
+      }
+
+      // Aggregate the metrics.
+      long timestamp = values.getTimestamp();
+      MetricsAggregator metricsAggregator = aggregated.computeIfAbsent(
+          new MetricsMapKey(values.getTags(),
+              getRoundedTimestamp(values.getTimestamp(), offset)),
+          (t) -> new MetricsAggregator());
+      for (MetricValue metric : values.getMetrics()) {
+        metricsAggregator.addMetric(metric, timestamp);
+      }
+    });
+
+    return aggregated.entrySet().stream().map((entry) -> entry.getValue()
+        .getAggregateMetricValues(entry.getKey().getTags(),
+            entry.getKey().getRoundedTimestamp())).map((metricValues) -> {
+      // Encode the metric value into a message.
+      encodeOutputStream.reset();
+      try {
+        recordWriter.encode(metricValues, encoder);
+      } catch (IOException e) {
+        throw new RuntimeException(
+            "Failed to encode metric to message payload.", e);
+      }
+      byte[] payload = encodeOutputStream.toByteArray();
+      return (Message) new Message() {
+        @Override
+        public String getId() {
+          // TopicRelayers keep track of the last message ID
+          // to fetch messages in the next iteration.
+          // Message IDs are generated by the messaging service. Return the last message ID before aggregation
+          // every time.
+          return messageId[0];
+        }
+
+        @Override
+        public byte[] getPayload() {
+          return payload;
+        }
+      };
+    }).iterator();
+  }
+
+  private long getRoundedTimestamp(long timestamp, long offset) {
+    return (Math.max(timestamp - offset, 0) / aggregationWindowSeconds)
```

Comment:
What I meant in my earlier comment was that the calculation seems like could be just `timestamp / windowSeconds * windowSeconds`? Why need the `offset`?

Reply:
![Aggregation optimization](https://user-images.githubusercontent.com/46515553/227282926-85750573-f8b8-4fb2-a82f-abf903c8fafd.jpg)
If you look at the above diagram, on average, half of the last aggregation window will be in the future. Say we are using a window of 2 secs, the last 1 second will lie in the last window and the last [t-3, t-1) seconds will lie in the second last window. We know that in most cases, all the metrics will lie in the last 2 seconds, unless a metrics publisher is lagging. 

This optimization brings brings back all timestamps by an offset so that the last window contains maximum values.